addons/mass_mailing/models/mailing_list.py

```diff
@@ -136,3 +192,109 @@ def action_merge(self, src_lists, archive):
+    # UTILITY
+    # ------------------------------------------------------
+
+    def _fetch_contact_nbr_counts(self):
+        """ Compute number of contacts matching various conditions.
+        (see '_get_contact_nbr_select_fields' for details)
+
+        Will return a dict under the form:
+        {
+            42: { # 42 being the mailing list ID
+                'contact_nbr': 52,
+                'contact_nbr_email': 35,
+                'contact_nbr_opt_out': 5,
+                'contact_nbr_blacklisted': 2
+            },
+            ...
+        } """
+
+        res = []
+        if self.ids:
+            self.env.cr.execute(f'''
+                SELECT
+                    list_id AS mailing_list_id,
+                    {','.join(self._get_contact_nbr_select_fields().values())}
+                FROM
+                    mailing_contact_list_rel r
+                    LEFT JOIN mailing_contact c ON (r.contact_id=c.id)
+                    {self._get_blacklist_join()}
+                WHERE list_id IN %s
+                GROUP BY
+                    list_id;
+            ''', (tuple(self.ids), ))
+            res = self.env.cr.dictfetchall()
+
+        contact_nbr_counts = {}
+        for res_item in res:
+            mailing_list_id = res_item.pop('mailing_list_id')
+            contact_nbr_counts[mailing_list_id] = res_item
+
+        for mass_mailing in self:
+            # adds default 0 values for ids that don't have statistics
+            if mass_mailing.id not in contact_nbr_counts:
+                contact_nbr_counts[mass_mailing.id] = {
+                    field: 0
+                    for field in self._get_contact_nbr_select_fields().keys()
+                }
+
+        return contact_nbr_counts
+
+    def _get_contact_nbr_select_fields(self):
+        """ Returns fields and SQL query select path in a dictionnary.
+        This is done to be easily overridable in subsequent modules.
+
+        - contact_nbr:              all contacts
+        - contact_nbr_email:        all valid emails
+        - contact_nbr_opt_out:      all opted-out contacts
+        - contact_nbr_blacklisted:  all blacklisted contacts
+
+        The blacklisted condition will be overridden in mass_mailing_sms so that blacklisted
+        statistics also take the sms blacklist into account."""
+
+        return {
+            'contact_nbr': 'COUNT(*) AS contact_nbr',
+            'contact_nbr_email': '''
+                SUM(CASE WHEN
+                        (c.email_normalized IS NOT NULL
+                        AND COALESCE(r.opt_out,FALSE) = FALSE
+                        AND bl.id IS NULL)
+                        THEN 1 ELSE 0 END) AS contact_nbr_email''',
+            'contact_nbr_opt_out': '''
+                SUM(CASE WHEN COALESCE(r.opt_out,FALSE) = TRUE
+                    THEN 1 ELSE 0 END) AS contact_nbr_opt_out''',
+            'contact_nbr_blacklisted': f'''SUM(CASE WHEN {self._get_blacklisted_condition()}
```

Comment:
As we have one override, isn't it possible to remove _get_blacklisted_condition and simply rewrite the contact_nbr_blacklisted entry in override ? It would reduce number of tool methods.